content/blog/semgrep-post-message/index.md

```diff
@@ -0,0 +1,175 @@
+
+```js
+// Inline function without origin check
+window.addEventListener("message", function(evt){
+  console.log('No origin check!');
+});
+
+// External function without origin check
+function receiveMessage(evt) {
+  console.log('No origin check!');
+}
+
+window.addEventListener("message", receiveMessage, false);
+
+// Inline arrow function without origin check
+window.addEventListener('message', (evt) => {
+  console.log('No origin check!');
+});
+
+// External arrow function without origin check
+const arrowHandler = (evt) => {
+    console.log('No origin check!');
+};
+
+window.addEventListener("message", arrowHandler, false);
+```
+
+Knowing that we need to match `inline` or `external` declarations gives us an idea that we will need to rely on the [patterns-either](https://semgrep.dev/docs/writing-rules/rule-syntax/#pattern-either) operator. 
+
+We might be tempted to write four different patterns for this scenario, but since Semgrep is semantic we only need to write **two**! This means that patterns matching and old `function` declaration will also match code using `arrow functions`. 
+
+It's worth repeating this:
+
+> Since Semgrep is semantic, patterns matching and old `function` declaration will also match code using `arrow functions`
+
+ Now, which patterns do we consider safe for origin checking?
+
+## Safe patterns
+
+If the function specified above contains any of the following patterns we **should not** flag them:
+
+```js
+if (evt.origin == "http://example.com") { ... }
+
+if (evt.origin === "http://example.com") { ... }
+
+if (evt.origin != "http://example.com") { ... }
+
+if (evt.origin !== "http://example.com") { ... }
+
+if (someRegex.test(evt.origin)) { ... }
+```
+<br />
+We have our spec, we know what to flag and more importantly, what not to. Time to start working on our rule!
+
+## Writing our rule
+
+Let's take care of the `inline` functions first since they don't need external context:
+
+#### Inline functions
+
+```yaml
+patterns:
+- pattern: |
+  window.addEventListener('message', $FUNC, ...)
+```
+
+`$FUNC` will capture the inline function into a [metavariable](https://semgrep.dev/docs/writing-rules/pattern-syntax/#metavariables). Now we can use the handy [metavariable-pattern](https://semgrep.dev/docs/writing-rules/rule-syntax/#metavariable-pattern) to exclude our **safe patterns**.
+
+```yaml
+patterns:
+- pattern: |
+    window.addEventListener('message', $FUNC, ...)
+- metavariable-pattern:
+    metavariable: $FUNC
+    patterns:
+      - pattern: |
+          function($OBJ) { ... }
+      - pattern-not: |
+          function($OBJ){ if ($OBJ.origin == $X) { ... } }
+      - pattern-not: |
+          function($OBJ){ if ($OBJ.origin === $X) { ... } }
+      - pattern-not: |
+          function($OBJ){ if ($OBJ.origin != $X) { ... } }
+      - pattern-not: |
+          function($OBJ){ if ($OBJ.origin !== $X) { ... } }
+      - pattern-not: |
+          function($OBJ){ if ($REGEX.test($OBJ.origin)) { ... } }
```

Comment:
Random idea: as an AppSec engineer you might not know where all of the usages of `postMessage` are and what they look like. 

1. So the first step is trying to understand your environment, so you just look for `window.addEventListener(... $MSG, $FUNC, ...)`, and manually review $MSG and $FUNC in the results.
2. You find some interesting patterns in some of the handlers, maybe checks on the `origin` like in this blog post, so you grab all of the conditionals `if ($COND) {...}` that contain `$OBJ.origin`, and review all of the `$COND`s - are there trends, outliers?

Ideally at the end of the day there's some `security_approved_origin_validator($OBJ.origin)`, that you can ensure is used, vs one-off dev implementation, but that's longer term strategic stuff that takes more collaboration and coordination.

Anywho, random idea about trying to iteratively understand your environment with Semgrep and then narrow down the focus based on what you see

Reply:
💯 We actually had a similar event with `dangerouslySetInnerHTML`. In some cases we couldn't avoid it, so we catalogued these scenarios and crafted a `sanitize` function to be used in conjunction with it. After that we crafted a rule that ensured that an instance of `dangerouslySetInnerHTML` without the use of our custom sanitizer would be flagged.